Jdp client port prototype

grgustaf opened the pull request.

grgustaf committed:
initial working version of pulling in lots of the HTML client code

Ported to typescript, pulled out some things into classes, etc.

No unit tests yet and some commented-out code and other garbage to be
cleaned up, but wanted to save these signs of life.

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,404 @@
+import { Breakpoint, ParsedFunction } from './breakpoint';
+import { PROTOCOL } from './jrs-protocol-constants';
+import { FunctionMap, ParsedSource } from './parsed-source';
+import { cesu8ToString } from './string-utils';
```
I'm missing these new files. Did you forget to add them?

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,404 @@
+
+  connect() {
+    if (this.socket) {
+      this.disconnect();
```
Still think it makes more sense to `return Promise.resolve();` here instead of disconnecting.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,404 @@
+  }
+
+  getBreakpoint(breakpointData: Array<number>) {
+    const func = this.functions[breakpointData[0]];
```
I like your idea of having this `functions` map, but why not do it much easier in the process?
I would try to create a map where the key is the first byte of `message` in `onMessage`.
That way you can make the code a bit cleaner and replace this big switch:

```
  switch (message[0]) { 
```

martijnthe replied:
I just read your questions in IRC. Sorry for the delay...

For some reason TS doesn't like using `PROTOCOL.SERVER.XYZ` as key in an `interface`. I'm guessing because property values can be dynamically created at runtime (using JS' `get`ter functionality).

However, it does seem to be OK with simple `const` numbers -- I've not tried, but I'm pretty confident it's also OK when these plain `const`s are imported from another file.

So this worked for me:

```
const JERRY_DEBUGGER_BREAKPOINT_HIT = 16; // change to import from ...
const JERRY_DEBUGGER_EXCEPTION_HIT = 23413243;
const JERRY_DEBUGGER_MEMSTATS = 1231;  

// jerry-dbg-parser.ts

type BytecodeOffset = number;

interface JerryMessageMemstats {
    allocatedBytes: number;
    byteCodeBytes: number;
    stringBytes: number;
    objectBytes: number;
    propertyBytes: number;
}

interface JerryMessageBreakpointHit {
    offset: BytecodeOffset;
}

interface JerryMessageExceptionHit {
    offset: BytecodeOffset;
}

interface ParserFunctionMap {
    [JERRY_DEBUGGER_MEMSTATS]: (Uint8Array) => JerryMessageMemstats;
    [JERRY_DEBUGGER_BREAKPOINT_HIT]: (Uint8Array) => JerryMessageBreakpointHit;
    [JERRY_DEBUGGER_EXCEPTION_HIT]: (Uint8Array) => JerryMessageExceptionHit;
}
```

grgustaf replied:
I've gone back and forth trying to understand the point of that FunctionMap you've outlined. I guess to replace the switch statement, you intend that you would indirect through the map to get to a function that will handle each message. One issue is what happens if there is a bogus protocol message with an unexpected 'type'; I guess this can be handled by an exception that there is no such key in the map.

But then if each function in the map returns a different datatype like this, how am I supposed to deal with the return value? So currently I'm having these return void and assume that each function will handle its message type and if useful call a delegate function.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,404 @@
+          }
+        }
+
+        this.lastBreakpointHit = breakpoint;
```
~Could you cleanly separate the message parsing vs debugger client state tracking?~

EDIT: the protocol is very state-ful, so I think it makes sense to combine protocol state (such as the function list state) and the parsing in one file.

That said, I would split off the connection stuff. Much easier to unit test that way if the protocol code file doesn't drag / require the connection code.

Personally I would use the delegate pattern there (I'm assuming there's only one thing interested in hearing about the parsed messages; the client itself). The protocol parser part would have a "delegate" property: an object that would be called back for each parsed message object. Something along these lines:

```
interface JerryProtocolParserDelegate {
  // probably makes sense to not require them but make them optional, for testing but also if someone wants to just re-use the parser.
  onDebuggerExceptionHit?(message: JerryMessageExceptionHit): void;
  onDebuggerMemstats?(message: JerryMessageMemstats): void;
  // .. etc
}

class JerryProtocolParser {
  public delegate: JerryProtocolParserDelegate;
  parseData(data: ArrayBuffer);
}
```

The client would instantiate a parser, assign itself as the delegate, implement the callbacks it's interested in and call `parseData` with the data the client gets in `onMessage`.

grgustaf replied:
I thought I understood this, but TS seems to be complaining if it I try to use the client as the delegate, because it implements more than just what is defined in the Delegate interface. It has to allow multiple interface inheritance right? So I'm probably just doing something wrong. Anyway, just recording my puzzle points in case you have time to comment before I find an answer.

grgustaf replied:
OK this turned out to just need at least one method in the interface to be actually defined.

martijnthe replied:
OK, glad you figured it out.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,404 @@
+    return {
+      breakpoint: func.offsets[nearestOffset],
+      at: false,
+    };
```
I think it would be nice to define `interface`s for each of the JrS debugger messages and use then in the return type of each of the parser functions.

grgustaf replied:
Now this makes sense I think to have an interface defined for the 'options' to each delegate callback. I assume common sets of arguments could reuse the same interface: split them out later if something changes.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,404 @@
+    console.log('Abort connection: ' + message);
+    throw new Error(message);
+  }
+
```
I like the idea of using python-esk `struct` formatting.
I'd put it in a separate file though. Also add some tests please.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,404 @@
+    let value;
+
+    if (offset + this.getFormatSize(format) > message.byteLength) {
+      this.abortConnection('received message too short');
```
EDIT: Actually, perhaps that is not true because we're using a WebSocket client here and not a raw socket like in the python script. I think WS protocol should already handle framing.

~The `jerry-client-ws.py` python implementation (see `get_message`) suggests to me that it's possible to get partial messages. In other words, the parser should be able to gather and "reassemble" partial messages.~

martijnthe commented on jerry-debugger/src/lib/parsed-source.ts:
```diff
@@ -0,0 +1,178 @@
+  private functions: FunctionMap = {};
+  private lastScriptID: number = 0;
+
+  constructor(debugClient: JerryDebuggerClient) {
```
Couple high level comments:

I'd try to avoid node-isms in the parser / protocol code. Relying on `EventEmitter` (a Node API) will make it trickier to make this usable in a browser environment (i.e. the HTML client).
AFAICT there's only one interested party in the `scriptParsed` events, so this mechanism might as well be replaced with the delegate pattern.

Something along these lines:

```

export interface ScriptParsedResult {
  scriptId: number;
  name: string;
  endLine: number;
  endColumn: number,
  hasSourceURL: boolean;
}

export interface ParsedSourceDelegate {
  onScriptParsed(result: ScriptParsedResult): void;
}

export class ParsedSource {
 ...
  // the client would be passed here still, but by having a smaller interface (`ParsedSourceDelegate`) it will
  // make unit testing this simpler because we don't have to mock the whole client, just the delegate methods.
 constructor(public delegate: ParsedSourceDelegate) {
     ...
 }

 ..
 // instead of this.emit('...', obj), we'll do this.delegate.onScriptParsed(obj);

}

```

grgustaf committed:
separate websocket code from message parsing code

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf commented:
OK, I did replace the switch with a function map at least, in this version. But it came out rather differently than you sketched out so may need more feedback.

I also separated the WebSocket code form the protocol implementation. I have started the delegation but not sure if the socket-owning JerryDebuggerClient class is really what we mean to delegate to. Maybe the socket part should be further split out from it.

I'll continue with pulling out the decoding stuff, adding delegation calls, unit tests...

grgustaf committed:
fix lint warnings

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf commented:
@martijnthe if you have a chance, take a look at the commit message of my latest commit here, i describe what has me blocked right now.

grgustaf commented:
Resolved the promise type error I was having and folded the change into the top commit.

grgustaf committed:
move decode functions out to utils.ts (renamed from string-utils.ts)

Also, add unit tests and fix a math bug revealed by a unit test.

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
add initial cesu8ToString tests

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

Signed-off-by: Geoff Gustafson <geoff@linux.intel.com>

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
add unit tests for concatUint8Arrays

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
Rename protocol-parser.ts to protocol-handler.ts

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf commented:
@martijnthe, I realize I haven't yet figured out when to throw an error vs. call the delegate's onError() if you care to comment on that.

grgustaf commented:
I'm about to submit a patch doing some testing on ProtocolHandler. I realize for one thing my onFunctions are public, not sure if that's right - but it is convenient in that I can write unit tests for them. Also, most of my data is private so there's not much I can check from unit tests, apart from whether delegate functions were called. So that could be worth comment. I'm not sure how valuable this testing is in this next patch or how much is missing; maybe things need to be further restructured to be more testable, but I could use another round of comment on that.

grgustaf committed:
add partial unit tests for protocol handler

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
rename protocol-parser test

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
send initial scriptParsed event

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
fix bug expecting only windows-style newlines

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

martijnthe commented on jerry-debugger/src/lib/utils.ts:
```diff
@@ -0,0 +1,148 @@
+  return result;
+}
+
+export function cesu8ToString(array: Uint8Array | undefined) {
```
Can you add the return type `: string` ?

martijnthe commented on jerry-debugger/src/lib/__tests__/utils.test.ts:
```diff
@@ -0,0 +1,191 @@
+  });
+
+  // TODO: test the part of CESU-8 incompatible with UTF-8; however, the existing
+  //   debugger client code didn't actually support this
```
I suspect no additional code is needed in `cesu8ToString`. The difference between CESU-8 and UTF-8 is that CESU-8 uses "surrogate pairs" to encode code points that fall outside of the basic multilingual plane (which would be encoded into 4-bytes of UTF-8 data otherwise). "Surrogate pairs" come from UTF-16, which not coincidentally ECMA/JS also uses its "language-native encoding". AFAIK, surrogate pairs is the (only) way to represent those code points outside of the basic multilingual plane in a JS string.

A simple test case would be to add convert a string with an emoji (using 😂 here as an example):

UTF-8 data:
`0xF0, 0x9F, 0x98, 0x82`

UTF-16 surrogate code points:
`0xD83D, 0xDE02`

I just tried in Node.js:
```
> '😂'.charCodeAt(0).toString(16)
'd83d'
> '😂'.charCodeAt(1).toString(16)
'de02'

> String.fromCharCode(0xD83D) + String.fromCharCode(0xDE02)
'😂'

> Buffer.from('😂')
<Buffer f0 9f 98 82>  // utf-8 data
```

grgustaf replied:
I attempted to just add this emoji test, if I understood it. However, according to [wikipedia](https://en.wikipedia.org/wiki/CESU-8), there are six-byte CESU8 encodings, when the UTF-8 encoding of a surrogate point falls in the 3-byte range, and that's what I think might not be handled.

I don't fully understand it though.

grgustaf replied:
Oops my test wasn't actually working, backing it out for now.

martijnthe replied:
Added a test for it:
https://github.com/jerryscript-project/jerryscript/pull/2205/commits/44493c71eb26782e4b39c0388cf730ac1068bbcf

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+import { ByteConfig, cesu8ToString, concatUint8Arrays, decodeMessage } from './utils';
+
+// expected JerryScript debugger protocol version
+const JERRY_DEBUGGER_VERSION = 1;
```
Move into `jrs-protocol-constants.ts`?

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+  sourceName?: string;
+  lines: Array<number>;
+  offsets: Array<number>;
+  byteCodeCP?: number;
```
What does CP stand for? Maybe pick a more obvious name?

grgustaf replied:
Compressed pointer... I'll at least change the type to an explicit typedef (as you suggested elsewhere) but I think this is a useful abbreviation as that's a rather unhelpful mouthful. :)

martijnthe commented on jerry-debugger/src/lib/utils.ts:
```diff
@@ -0,0 +1,148 @@
+}
+
+// Concat the two arrays. The first byte (opcode) of nextArray is ignored.
+export function concatUint8Arrays(baseArray: Uint8Array | undefined, nextArray: Uint8Array) {
```
nit: the name doesn't reveal it's doing more than plain concatenation.
Perhaps `assembleChunks` or something like that?

grgustaf replied:
Yeah, agreed... rather odd. Inherited from existing code. :)

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+    }];
+  }
+
+  // for the temporary mollification of lint
```
You can add a magic comment above the member to avoid doing this, i.e.
```
/* tslint:disable-next-line:no-unused-variable */
private maxMessageSize: number = 0;
```

After the 2nd colon, you'd put the rule that you want to ignore.

See https://palantir.github.io/tslint/usage/rule-flags/

grgustaf replied:
OK, that does seem to work for lint, but the test suites still don't run because of this type of error. So I'm leaving it in but changing the comment.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+      source: '',
+      lines: [],
+      offsets: [],
+    }];
```
Why can't we init with an empty array?

grgustaf replied:
This object represents the entire source file so it knows it's done parsing the source file when this last object is popped off the stack. I'm sure there's another way to do it but I'm just mimicking the existing working code for now.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+    return decodeMessage(this.byteConfig, format, message, offset);
+  }
+
+  onConfiguration(data: Uint8Array) {
```
> I realize for one thing my onFunctions are public, not sure if that's right - but it is convenient in that I can write unit tests for them. 

Re. public: we've been using the idea of "cracker" subclasses in our code. The idea is simple: use `protected` and then declare a "cracker" subclass in the test that changes the visibility from `protected` to `public`.

```
// impl.ts
class JerryDebugProtocolHandler {
  protected onConfiguration(data: Uint8Array) { ... }
  protected onByteCodeCP(data: Uint8Array) { ... }
  // ...etc
}

// ....test.ts:
class JerryDebugProtocolHandlerCracker extends JerryDebugProtocolHandler {
  public onConfiguration(data: Uint8Array);
  public onByteCodeCP(data: Uint8Array);
  // ...etc
}

// using it -- just cast to the cracker type:
const handler = new JerryDebugProtocolHandler(delegate) as JerryDebugProtocolHandlerCracker;
handler.onConfiguration; // this is legal now

```

grgustaf replied:
OK, I'll try to keep that in mind when I notice  this again but I'm not going to scour the code for it right this moment.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+  private functionName?: Uint8Array;
+  private functions: FunctionMap = {};
+  private lastScriptID: number = 0;
+
```
> Also, most of my data is private so there's not much I can check from unit tests, apart from whether delegate functions were called.

I think this is fine. Personally, I think the less the tests know about the implementation internals, the better.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+    }
+  }
+
+  parseData(data: ArrayBuffer) {
```
Why take `ArrayBuffer` and not `Uint8Array` here?
I would push the conversion to `Uint8Array` out one layer (into what is currently called `debugger-client.ts`).
`ArrayBuffer` is a bit cumbersome to work with because you always have to create a view before you can do anything with it.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+}
+
+export interface JerryDebugProtocolDelegate {
+  onError?(message: string): void;
```
Can you add an error code (i.e. enum) as well?

grgustaf replied:
Faked it for now, can flesh that out later.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+  }
+
+  onSourceCode(data: Uint8Array) {
+    console.log('[Source Code]');
```
Maybe create a new `logging.ts` that allows turning logs on/off per log level?

grgustaf replied:
OK, I've moved a couple of these things out to issues since I'd prefer to do this later and prioritize core progress I can feel better about. :)

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+
+  onConfiguration(data: Uint8Array) {
+    console.log('[Configuration]');
+    if (data.length !== 5) {
```
Maybe do `<` instead?
In theory, if the configuration message changed, the version should have been bumped.
I think it's safe to assume the message won't be *shorter* in newer versions.
I think it would be more friendly/useful to hit the "incorrect target debugger version" error than to get "configuration message wrong size".

(Too bad the version isn't the first byte BTW)

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+    for (let i = 1; i < data.byteLength; i += 4) {
+      array.push(this.decodeMessage('I', data, i)[0]);
+    }
+    return;
```
`return;` not needed

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+      array = this.stack[this.stack.length - 1].lines;
+    } else {
+      array = this.stack[this.stack.length - 1].offsets;
+    }
```
Maybe create a local `const stackFrame = this.stack[this.stack.length - 1]` to avoid the duplication + make it easier to read?

martijnthe commented on jerry-debugger/src/lib/__tests__/protocol-parser.test.ts:
```diff
@@ -0,0 +1,116 @@
+    expect(delegate.onScriptParsed).toHaveBeenCalledTimes(1);
+    const data = delegate.onScriptParsed.mock.calls[0][0];
+    expect(data.endLine).toEqual(1);
+  });
```
If I read it correctly, in both cases, the concatenation/reassembly logic doesn't have any previous data.
Perhaps add a test cases that covers this? i.e receiving a `JERRY_DEBUGGER_SOURCE_CODE` and then a `JERRY_DEBUGGER_SOURCE_CODE_END` message?

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+    console.log('[Source Code Name]');
+    this.sourceNameData = concatUint8Arrays(this.sourceNameData, data);
+    if (data[0] === SP.JERRY_DEBUGGER_SOURCE_CODE_NAME_END) {
+      this.sourceName = cesu8ToString(this.sourceNameData);
```
Do you still need `sourceNameData` after this point?
If not perhaps set to `undefined`?

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+    console.log('[Source Code]');
+    this.sourceData = concatUint8Arrays(this.sourceData, data);
+    if (data[0] === SP.JERRY_DEBUGGER_SOURCE_CODE_END) {
+      this.source = cesu8ToString(this.sourceData);
```
Same here, do you still need `sourceData` after this point? If not, clean up and set to `undefined`?

martijnthe commented on jerry-debugger/src/lib/breakpoint.ts:
```diff
@@ -0,0 +1,96 @@
+                + this.func.line
+                + ', col:'
+                + this.func.column
+                + ')';
```
I'd use template strings here instead:

```
result += ` (in ${this.func.name || 'function'}() at line:${this.func.line}, col: ${this.func.column})`;
```

martijnthe commented on jerry-debugger/src/lib/breakpoint.ts:
```diff
@@ -0,0 +1,96 @@
+  func: ParsedFunction;
+  line: number;
+  offset: number;
+  activeIndex?: number;
```
In what case is this unset? It looks like it's always initialized to `-1`?
Remove the `?` perhaps?

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+    if (data[0] === SP.JERRY_DEBUGGER_SOURCE_CODE_NAME_END) {
+      this.sourceName = cesu8ToString(this.sourceNameData);
+      // assumption: this will be completed before source and included in the
+      //   onScriptParsed delegate function called in onSourceCode
```
Maybe test the assumption here at runtime and `abort` in case it fails?

grgustaf replied:
Recorded with a TODO

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -23,27 +25,33 @@ export interface JerryDebuggerOptions {
     this.host = options.host || DEFAULT_DEBUGGER_HOST;
     this.port = options.port || DEFAULT_DEBUGGER_PORT;
     this.verbose = options.verbose || false;
+    this.protocolHandler = new JerryDebugProtocolHandler(this);
```
When you rework this according to your design diagram, I'd probably inject the "transport" (websocket client) as well as the delegate when constructing the `JerryDebugProtocolHandler`. This makes it easier to test things as well as make it easier to use different kinds of transports as well.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+  offsets: Array<number>;
+  byteCodeCP?: number;
+  firstBreakpointLine?: number;
+  firstBreakpointOffset?: number;
```
I don't see these being used as part of `ParserStackFrame`.
And these fields are also part of `ParsedFunction`. Did you forget to remove them perhaps?

martijnthe commented on jerry-debugger/src/lib/breakpoint.ts:
```diff
@@ -0,0 +1,96 @@
+  readonly column: number;
+  readonly name: string;
+  readonly firstBreakpointLine: number;
+  readonly firstBreakpointOffset: number;
```
Wouldn't it make more sense to instead have
```
readonly firstBreakpoint: Breakpoint;
```

What will be the use of this BTW?

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+    this.functions[byteCodeCP] = func;
+    if (this.stack.length > 0) {
+      return;
+    }
```
I'm a bit fuzzy on why the `func` gets assigned `source` and `sourceName` if it was the last/bottom frame.
If other `func`s don't have these assigned, won't the `toString()` function for breakpoints in that func print `<unknown>` as sourceName?

grgustaf replied:
I'm afraid all these last few questions I'm also a bit fuzzy on as they just come from the way that the original HTML debugger was written. But they did serve some purpose. I'd prefer to progress further in the protocol and deal with these things when I get up to working on the functionality that requires them.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -0,0 +1,251 @@
+      line: position[0],
+      column: position[1],
+      name: cesu8ToString(this.functionName),
+      source: this.source,
```
Later on, `this.source` is being split using `split(/\r\n[\r\n]/)`.
In `onSourceCode`, it's also being split.
I wonder if it makes sense to split it once and use the result everywhere instead (and ditch the non-splitted string.)

grgustaf committed:
restructure to for sanity removing circular dependency

Also, add some support for breakpoint hit

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
add type info to utils based on Martijn's comments

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf commented:
@martijnthe, I'm getting 'test suite failed to run' for the cli tests and haven't been able to debug it if you care to take a look. There doesn't seem to be any more helpful error message.

SCRATCH that, this was the test suite giving up because of a stupid lint-type error.

grgustaf committed:
respond to feedback and fix most lint warnings

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com
